### PR TITLE
Add a linear dm device on the top of a nvme device

### PR DESCRIPTION
**Pull request by derekbit**
longhorn/longhorn#6001

### CODE REVIEW

File: pkg/nvme/initiator.go [hunk 1]
```diff
@@ -271,3 +334,96 @@ func (i *Initiator) removeEndpoint() error {
 
 	return nil
 }
+
+func (i *Initiator) removeLinearDmDevice() error {
+	devPath := fmt.Sprintf("/dev/mapper/%s", i.Name)
```
Comment (shuo-wu) [inline pkg/nvme/initiator.go:339]: Add a function `GetDMDevicePath`

File: pkg/nvme/initiator.go [hunk 2]
```diff
@@ -271,3 +334,96 @@ func (i *Initiator) removeEndpoint() error {
...
+	if _, err := os.Stat(devPath); err != nil {
```
Comment (shuo-wu) [inline pkg/nvme/initiator.go:340]: Can we validate the error via `os.IsNotExist`?

File: pkg/util/device.go [hunk 3]
```diff
@@ -14,15 +16,27 @@ import (
...
+type BlockDevice struct {
+	Name   string `json:"name"`
+	Major  int    `json:"maj"`
+	Minor  int    `json:"min"`
+	MajMin string `json:"maj:min"`
+}
+
+type BlockDevices struct {
+	Devices []BlockDevice `json:"blockdevices"`
+}
```
Comment (shuo-wu) [inline pkg/util/device.go:32]: NIT: Do we need to define such a struct?

File: pkg/util/device.go [hunk 4]
```diff
@@ -14,15 +16,27 @@ import (
 )
 
 const (
-	LSBLKBinary = "lsblk"
+	lsblkBinary    = "lsblk"
+	blockdevBinary = "blockdev"
 )
 
+type BlockDevice struct {
```
Comment (shuo-wu) [inline pkg/util/device.go:23]: NIT: Add "path" for this struct

File: pkg/nvme/initiator.go [hunk 5]
```diff
@@ -271,3 +334,96 @@ func (i *Initiator) removeEndpoint() error {
...
+	return util.DmsetupSuspend(i.Name, i.executor)
+}
+
+func (i *Initiator) resumeLinearDmDevice() error {
+	logrus.Infof("Resuming linear dm device %s", i.Name)
+
+	return util.DmsetupResume(i.Name, i.executor)
+}
+
+func (i *Initiator) reloadLinearDmDevice() error {
```
Comment (shuo-wu) [inline pkg/nvme/initiator.go:408]: When will we need to use this? Online expansion? Do we need to apply suspend before invoking this?

**Review by shuo-wu — COMMENTED** [review]
In general LGTM. 

I have 2 questions about this PR:
1. When will we need to add `--ipc` for ns execution?
2. Will the dmsetup related util functions be moved to https://github.com/longhorn/go-common-libs in the future?

**Comment by derekbit** [comment]
> In general LGTM.
> 
> I have 2 questions about this PR:
> 
> 1. When will we need to add `--ipc` for ns execution?

Enter host ipc namespace for dmsetup command. If the ipc namespace is not used, dmsetup will hang and not return even though the appropriate action will be executed in the background.

> 2. Will the dmsetup related util functions be moved to https://github.com/longhorn/go-common-libs in the future?

Yes, I would like to move it to go-common-libs in the future.

File: pkg/nvme/initiator.go (hunk 5 above)
Comment (derekbit) [reply pkg/nvme/initiator.go:408]: No. Thie is not used for now. We can remove it and add it later when working on online exapnsion.

> Do we need to apply suspend before invoking this?

Yes, need to suspend first.

File: pkg/util/device.go (hunk 3 above)
Comment (derekbit) [reply pkg/util/device.go:32]: Yes, it maps to the output `lsblk -l -J -n -o MAJ:MIN <path`. Can we avoid the definition?

File: pkg/nvme/initiator.go [hunk 6]
```diff
@@ -70,9 +77,32 @@ func NewInitiator(name, subsystemNQN, hostProc string) (*Initiator, error) {
...
+	defer func() {
+		if err != nil {
+			i.logger.WithError(err).Errorf("Failed to start NVMe initiator %s", i.Name)
+		}
+	}()
+
 	if transportAddress == "" || transportServiceID == "" {
-		return fmt.Errorf("empty TransportAddress %s and TransportServiceID %s for initiator %s start", transportAddress, transportServiceID, i.Name)
+		return fmt.Errorf("invalid TransportAddress %s and TransportServiceID %s for initiator %s start", transportAddress, transportServiceID, i.Name)
 	}
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/nvme/initiator.go:106]: Do you mean we need to blindly clean up the leftover/stale dm device first?

File: pkg/util/device.go (hunk 4 above)
Comment (derekbit) [reply pkg/util/device.go:23]: What's the purpose of the `path`?

**Comment by derekbit** [comment]
Close the PR and handled it in v2 engine upgrade together.

**Pull request closed without merging**